app/services/authentication/providers.rb

```diff
@@ -48,5 +48,11 @@ def self.enabled
     def self.enabled?(provider_name)
       enabled.include?(provider_name.to_sym)
     end
+
+    def self.username_fields
+      Authentication::Providers::Provider.subclasses.map do |subclass|
+        subclass.const_get("USERNAME_FIELD").to_sym
```

Comment:
I think we can provide reasonably defaults in the base class for `user_username_field`, `user_created_at_field` and `official_name`:

```ruby
def self.official_name
  name
end

def self.user_username_field
  "#{name.underscore}_username".to_sym
end 

def self.user_created_at_field
  "#{name.underscore}_created_at".to_sym
end

def self.settings_url
  raise SubclassResponsibility
end
```

Then we can get rid of the constants and stop defining these methods in every subclass, instead only override them where necessary (e.g. "GitHub" as `official_name`). If you're worried about performance you can cache them in instance variables, but I think most of these will be accessed mostly during initial app load for metaprogramming purposes. 

If we do this, this block would become:

```ruby
Authentication::Providers::Provider.subclasses.map(&:user_username_field).sort
```

Reply:
done! :)